### ISSUE TITLE
Issues in browser environments

### ISSUE THREAD

**Issue by RMHonor** [MyPureCloud/genesys-cloud-webrtc-sdk#576]
When trying to run the WebRTC SDK in the browser (Chrome 91) we hit issues where the package was looking for global variables that would only exist in a NodeJS environment (`process` and `Buffer`).

To fix this, we added the following to our webpack config:

```typescript
new webpack.ProvidePlugin({
  Buffer: ["buffer", "Buffer"],
}),
new webpack.DefinePlugin({
  process: {
    browser: true,
    nextTick: () => null,
  },
}),
```

We also needed to install the `stanza` package.

Is there a better solution for this that doesn't involve polyfilling?

SDK version: 6.0.1

**Comment by jensengar** [comment]
Can you show me your webpack config?

**Comment by RMHonor** [comment]
@jensengar Ours is quite complex, I'll just share the key files that will probably affect it:

Plugins:
```typescript
const plugins = [
    new webpack.ProvidePlugin({
      Buffer: ["buffer", "Buffer"],
    }),
    new webpack.DefinePlugin({
      process: {
        browser: true,
        nextTick: () => null,
      },
    }),
    new ProgressBarPlugin(),

    /**
     * Provide Env vars
     */
    new EnvironmentPlugin(env),
    new DotEnv({
      safe: true, // load '.env.example' to verify
      allowEmptyValues: false,
      silent: false,
    }),

    /**
     * Provide HTML template
     */
    new HtmlWebpackPlugin({
      template: path.join(paths.root, "public", "index.ejs"),
      GIT_SHA: env.GIT_SHA,
      PUBLIC_URL: env.PUBLIC_URL,
    }),

    /**
     * Add assets to dist
     */
    new CopyWebpackPlugin({
      patterns: [
        {
          from: path.join(paths.root, "public"),
          to: path.join(paths.build),
        },
      ],
    }),
  ];

  if (env.analyze) {
    plugins.push(
      new BundleAnalyzerPlugin({
        openAnalyzer: true,
      }),
    );
  }

  if (env.isDevelopment()) {
    plugins.push(
      new webpack.HotModuleReplacementPlugin(),
      new ForkTsCheckerWebpackPlugin({
        typescript: {
          build: true,
        },
      }),
    );
  }
```

Root config:

```typescript
const createWebpackConfig = (args: WebpackArgs): Configuration => {
  const env = createWebpackEnv(args);
  const paths = createWebpackPaths(PATH_ROOT);

  return {
    mode: env.isProduction() ? "production" : "development",
    entry: paths.src,
    output: {
      chunkFilename: "chunk.[name].[contenthash].js",
      path: paths.build,
      filename: "[name].[contenthash].js",
      publicPath: "/",
    },
    resolve: {
      extensions: [".tsx", ".ts", ".js", "jsx", ".native.js"],
      alias: {
        "react-native$": "react-native-web",
        web: path.resolve(PATH_ROOT, "src/web"),
        assets: path.resolve(PATH_ROOT, "src/web/assets"),
        bindings: path.resolve(PATH_ROOT, "src/web/bindings"),
        components: path.resolve(PATH_ROOT, "src/web/components"),
        hocs: path.resolve(PATH_ROOT, "src/web/hocs"),
        hooks: path.resolve(PATH_ROOT, "src/web/hooks"),
        modules: path.resolve(PATH_ROOT, "src/web/modules"),
      },
    },
    module: {
      rules: [jsRule, graphqlRule, cssRule, scssRule, assetRule],
    },
    plugins: createWebpackPlugins(env, paths),
    devtool: env.isProduction() ? "source-map" : "eval-source-map",
    cache: env.isProduction() ? false : true,
    optimization: {
      minimize: env.isProduction(),
      splitChunks: {
        chunks: "all",
        minSize: 20000,
        minChunks: 1,
        maxAsyncRequests: 30,
        maxInitialRequests: 30,
        cacheGroups: {
          defaultVendors: {
            test: /[\\/]node_modules[\\/]/,
            priority: -10,
            reuseExistingChunk: true,
          },
          default: {
            minChunks: 2,
            priority: -20,
            reuseExistingChunk: true,
          },
        },
      },
    },
    ...(env.isDevelopment() && {
      devServer: {
        open: true,
        compress: true,
        contentBase: [paths.build, paths.src],
        hot: true,
        port: 3000,
        historyApiFallback: true,
        overlay: {
          warnings: true,
          errors: true,
        },
      },
    }),
  };
};
```

JS loader:

```typescript
const jsRule: RuleSetRule = {
  test: /\.(ts|js)x?$/,
  loader: "babel-loader",
  exclude: /node_modules\/(?!(react-navigation-tabs|react-native-calendars|@react-navigation|react-native-screens)\/).*/,
};
```

**Comment by jensengar** [comment]
@RMHonor I'm honestly not sure. Your config is pretty complex and I'm not familiar with a lot of it. We have a sample (simple) webpack config in the project that works without having to pull in extra stuff. I believe stanza.io has the ability to run in node envs as well as browser and the fact that it's complaining about `Buffer` seems to hint at the environment not being configured correctly.

You can find the webpack sample in test/test-pages/webpack. You script that triggers the webpack build is [Here](https://github.com/MyPureCloud/genesys-cloud-webrtc-sdk/blob/master/test/test-pages/build.sh)

**Comment by djhouseknecht** [comment]
@RMHonor we found the issue. We are pointing `module` to a `commonjs` build of the sdk which is probably why you are seeing those node variables to need polyfilling. The `stanza` dependency bubbles up from the `genesys-cloud-streaming-client` dependency so I may need to work my way up through that. I will let you know when I have a fix ready and I'll give you the beta build you can install and try.

**Comment by RMHonor** [comment]
Thanks for the find @djhouseknecht 

I've created a much simpler implementation repo for testing against: https://github.com/utilitywarehouse/genesys-webrtc-test

**Comment by djhouseknecht** [comment]
@RMHonor thanks for the repo! I'll take a look at that to help track down the fix and verify it works.

**Comment by djhouseknecht** [comment]
@RMHonor I have a fix up. Would you be able to pull down the branch build and see if your issue is resolved? 

``` sh
npm i genesys-cloud-webrtc-sdk@6.0.2-branch.PCM1679.2
```

**Comment by RMHonor** [comment]
@djhouseknecht that works perfectly! No changes necessary in our build to get it to work 👌 

Is this likely to end up in master then?

**Comment by djhouseknecht** [comment]
Yep, opened #580 to merge into master and then we will get a release out of 6.0.2 which will contain the fix

**Comment by djhouseknecht** [comment]
Sorry for the delay on the PR. I thought I broken something with the imports, but it turns out that was an existing problem. This will go out in v6.1.0. We will work on getting that version released after #583 merges. Should be by the end of the week if approvals go through.
